### PR TITLE
Implement auth ui for SIWE

### PR DESCRIPTION
**Pull request by hochiw**
Screenshot 2022-09-07 at 3 54 04 PM
Screenshot 2022-09-07 at 3 54 12 PM
Screenshot 2022-09-07 at 5 09 32 PM

Note: Had to use a specific commit for the jazzicon package since the current npm version (2.0.0) has a vulnerability which was fixed in version 2.1.0 but is not on npm  

refs #2310

### CODE REVIEW

File: pkg/api/model/siwe.go [hunk 1]
```diff
@@ -1,14 +1,12 @@
...
 	Signature string `json:"signature"`
 }
 
 type SIWEVerifiedData struct {
-	Message          string  `json:"message"`
-	Signature        string  `json:"signature"`
-	EncodedPublicKey jwk.Key `json:"encoded_public_key"`
+	Message          string `json:"message"`
+	Signature        string `json:"signature"`
+	EncodedPublicKey string `json:"encoded_public_key"`
```
Comment (louischan-oursky) [inline pkg/api/model/siwe.go:11]: Add a newtype to represent the public key

```golang
type SIWEPublicKey string

func NewSIWEPublicKey(k *ecdsa.PublicKey) SIWEPublicKey {
  return hex.EncodeToString(crypto.CompressPubkey(k))
}

func (k SIWEPublicKey) ECDSA() (*ecdsa.PublicKey, error) {
  ...
}
```

Test it.

File: pkg/lib/authn/identity/info.go [hunk 2]
```diff
@@ -75,9 +75,10 @@ func (i *Info) ToSpec() Spec {
 		return Spec{
 			Type: i.Type,
 			SIWE: &SIWESpec{
-				VerificationRequest: model.SIWEVerificationRequest{
-					Message:   i.SIWE.Data.Message,
-					Signature: i.SIWE.Data.Signature,
+				VerifiedData: model.SIWEVerifiedData{
```
Comment (louischan-oursky) [inline pkg/lib/authn/identity/info.go:78]: I guess we need to revert this commit as we are going to combine verification and form submission into a single step.

File: resources/authgear/templates/zh-HK/translation.json [hunk 3]
```diff
@@ -186,12 +186,16 @@
...
 
   "web3-provider-metamask": "MetaMask",
+  "web3-provider-metamask-install": "安裝 MetaMask",
 
   "confirm-web3-account-page-title": "帳號已連接",
   "confirm-web3-account-page-description": "已連接到 {provider}",
   "confirm-web3-account-page-next": "下一步",
   "confirm-web3-account-page-change-account": "更改帳號",
 
+  "missing-web3-wallet-page-title": "安裝加密銀包",
```
Comment (louischan-oursky) [inline resources/authgear/templates/zh-HK/translation.json:202]: Change the term.

File: resources/authgear/templates/en/translation.json [hunk 4]
```diff
@@ -78,6 +78,12 @@
   "error-invalid-selected-image": "The selected image is invalid. Try another one.",
   "error-no-passkey": "Passkey is not found on your device. {type, select, password{Sign in with password instead?} oob_otp_email{Sign in with one-time password instead?} oob_otp_sms{Sign in with one-time password instead?} other{Try another device?}}",
 
+  "error-metamask-user-rejected": "User rejected the request.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/en/translation.json:81]: "You rejected the request."

File: resources/authgear/templates/en/translation.json [hunk 5]
```diff
@@ -78,6 +78,12 @@
...
+  "error-metamask-unauthorized": "The requested account and/or method has not been authorized by the user.",
```
Comment (louischan-oursky) [inline resources/authgear/templates/en/translation.json:82]: "You did not authorize the operation"

File: resources/authgear/templates/zh-HK/translation.json [hunk 6]
```diff
@@ -78,6 +78,12 @@
   "error-invalid-selected-image": "你所選擇的圖片格式有誤。請嘗試使用另一張圖片。",
   "error-no-passkey": "在你的裝置沒有通行密匙。{type, select, password{請使用密碼登入。} oob_otp_email{請使用一次性密碼登入。} oob_otp_sms{請使用一次性密碼登入。} other{請嘗試使用另一個裝置。}}",
 
+  "error-metamask-user-rejected": "用戶拒絕了該要請求。",
```
Comment (louischan-oursky) [inline resources/authgear/templates/zh-HK/translation.json:81]: "你拒絕了請求。"

File: resources/authgear/templates/zh-HK/translation.json [hunk 7]
```diff
@@ -78,6 +78,12 @@
...
+  "error-metamask-unauthorized": "要求的賬號和/或請求方法沒有得到用戶的授權。",
```
Comment (louischan-oursky) [inline resources/authgear/templates/zh-HK/translation.json:82]: "你拒絕了授權。"

File: resources/authgear/templates/zh-HK/translation.json [hunk 8]
```diff
@@ -78,6 +78,12 @@
...
+  "error-metamask-unsupported-method": "無效的請求方式。",
+  "error-metamask-disconnected": "已從區塊鏈斷開連接。",
+  "error-failed-to-connect-wallet": "無法連接到電子銀包。",
```
Comment (louischan-oursky) [inline resources/authgear/templates/zh-HK/translation.json:85]: Omit 電子

File: resources/authgear/templates/zh-TW/translation.json [hunk 9]
```diff
@@ -78,6 +78,12 @@
   "error-invalid-selected-image": "你所選擇的圖片格式有誤。請嘗試使用另一張圖片。",
   "error-no-passkey": "在你的裝置沒有通行密鑰。{type, select, password{請使用密碼登入。} oob_otp_email{請使用一次性密碼登入。} oob_otp_sms{請使用一次性密碼登入。} other{請嘗試使用另一個裝置。}}",
 
+  "error-metamask-user-rejected": "用戶拒絕了該要請求。",
```
Comment (louischan-oursky) [inline resources/authgear/templates/zh-TW/translation.json:81]: Apply the same change from zh-HK

File: authui/package.json [hunk 10]
```diff
@@ -32,6 +32,7 @@
     "@hotwired/stimulus": "3.1.0",
     "@hotwired/turbo": "7.1.0",
     "@tabler/icons": "1.92.0",
+    "@metamask/jazzicon": "github:MetaMask/jazzicon#d923914fda6a8795f74c2e66134f73cd72070667",
```
[the remark this answers is not in the document]
Comment (hochiw) [reply authui/package.json:35]: `2.0.0` has a dependency with vulnerability and `2.1.0` is not uploaded to npm

File: resources/authgear/templates/en/web/login.html [hunk 11]
```diff
@@ -4,15 +4,25 @@
...
+		{{ $has_siwe = true }}
+		{{ end }}
 	{{ end }}
 	<div class="pane flex flex-col">
+		{{ if $has_siwe }}
+		<div class="flex flex-row">
+			<a class="not-a flex items-center justify-center flex-1 py-4 signin-signup-link secondary-txt current" href="/login" data-turbo-action="replace">{{ template "login-label" }}</a>
+			<a class="not-a flex items-center justify-center flex-1 py-4 signin-signup-link secondary-txt" href="/signup" data-turbo-action="replace">{{ template "signup-label" }}</a>
+		</div>
+		{{ end }}
```
[the remark this answers is not in the document]
Comment (hochiw) [reply resources/authgear/templates/en/web/login.html:25]: thought it would be weird if we show login id params in the url, if it's fine I can remove them

**Comment by hochiw** [comment]
@louischan-oursky updated :adore2x:

File: resources/authgear/templates/en/web/login.html (hunk 11 above)
[the remark this answers is not in the document]
Comment (louischan-oursky) [reply resources/authgear/templates/en/web/login.html]: Let's NOT duplicate for now.

**Review by louischan-oursky — COMMENTED** [review]
- Wrap any call to metamask with try-catch. Display the error
- Just discard the user cancel error. It is not an error we want to show.
- Can we detect that the user has switch their account in metamask? If yes, I think we need not show the change account button. People should be familiar with switching their account in metamask rather then click our change account button.
- Test if mobile metamask is supported.
- The link to install metamask should be opened in a new tab.

**Comment by hochiw** [comment]
> * Wrap any call to metamask with try-catch. Display the error
> 
>     * Just discard the user cancel error. It is not an error we want to show.
> 
>     * Can we detect that the user has switch their account in metamask? If yes, I think we need not show the change account button. People should be familiar with switching their account in metamask rather then click our change account button.
> 
>     * Test if mobile metamask is supported.
> 
>     * The link to install metamask should be opened in a new tab.

@louischan-oursky updated accordingly except for the mobile one

https://github.com/MetaMask/detect-provider/issues/57

seems it's unable to be connected for some of the people too, but when I try out the sandbox someone made

https://z59g1u.csb.app/

it seems to be working... any insight?

File: authui/src/web3.ts [hunk 12]
```diff
@@ -207,17 +220,31 @@ export class WalletConfirmationController extends Controller {
...
-      return;
+    if (isProviderMetaMask(this.provider?.provider)) {
+      this.provider!.provider.on("accountsChanged", () => this._getAccount());
     }
+  }
 
-    this._getAccount();
+  disconnect() {
+    if (isProviderMetaMask(this.provider?.provider)) {
+      this.provider!.provider.off("accountsChanged", () => this._getAccount());
```
Comment (louischan-oursky) [inline authui/src/web3.ts:246]: The on/off pair MUST be passed the same function. Here we always pass them a fresh arrow function. So the function is leaked.

Here is how to accommodate the on/off pattern correctly.
```
class Foobar {
  connect() {
    this.provider.provider.on("accountsChanged", this.onAccountsChanged);
  }
  disconnect() {
    this.provider.provider.off("accountsChanged", this.onAccountsChanged);
  }

  // This is very important to define onAccountsChanged as a property.
  // It is bound to this.
  // If it were defined as
  // async onAccountsChanged() { }
  // We need to bind it with this.onAccountsChanged.bind(this)
  // But .bind always return a fresh bound function,
  // so our problem will remain unfixed.
  onAccountsChanged = async () => {
  }
}
```

**Comment by hochiw** [comment]
@louischan-oursky updated again :adore2x:

**Pull request merged**